contracts/utils/ExpiringMetaCallForwarder.sol

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+import "./SelfMulticall.sol";
+
+/// @notice Contract that forwards expiring meta-calls to ERC2771 contracts
+/// that trust it
+/// @dev The term "meta-call" is used as a subset of "meta-tx" where `value`
+/// is zero and `to` is a contract.
+/// Adapted from OpenZeppelin's MinimalForwarder.sol.
+contract ExpiringMetaCallForwarder is EIP712, SelfMulticall {
+    using ECDSA for bytes32;
+    using Address for address;
+
+    struct ExpiringMetaCall {
+        address from;
+        address to;
+        bytes data;
+        uint256 expirationTimestamp;
+    }
+
+    /// @notice If the meta-call with hash is already executed
+    mapping(bytes32 => bool) public metaCallWithHashIsExecuted;
+
+    bytes32 private constant _TYPEHASH =
+        keccak256(
+            "ExpiringMetaCall(address from,address to,bytes data,uint256 expirationTimestamp)"
+        );
+
+    constructor() EIP712("ExpiringMetaCallForwarder", "1.0.0") {}
+
+    /// @notice Verifies the signature and executes the meta-call
+    /// @dev This implementation is intended for use-cases where both `from`
+    /// and the meta-call executor benefit from the meta-call being executed
+    /// directly (e.g., `from` wants to whitelist the executor but does not
+    /// have funds, so they give the executor a meta-call to execute on their
+    /// behalf). This means a lot of the security considerations that apply in
+    /// general relayer network implementations are irrelevant here.
+    /// @param metaCall Meta-call
+    /// @param signature Meta-call hash signed by `from`
+    /// @return returndata Returndata
+    function execute(
+        ExpiringMetaCall calldata metaCall,
+        bytes calldata signature
+    ) external returns (bytes memory returndata) {
+        bytes32 metaCallHash = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    _TYPEHASH,
+                    metaCall.from,
+                    metaCall.to,
+                    keccak256(metaCall.data),
+                    metaCall.expirationTimestamp
+                )
+            )
+        );
+        require(
+            !metaCallWithHashIsExecuted[metaCallHash],
+            "Meta-call already executed"
+        );
+        require(
+            metaCall.expirationTimestamp > block.timestamp,
+            "Meta-call expired"
+        );
+        require(
+            metaCallHash.recover(signature) == metaCall.from,
+            "Invalid signature"
+        );
+        metaCallWithHashIsExecuted[metaCallHash] = true;
+        returndata = metaCall.to.functionCall(
+            abi.encodePacked(metaCall.data, metaCall.from)
+        );
```

Comment:
So this is going to work just fine for whitelisting a requester but if this contract is going to be used for other purposes, shouldn't it also forward `value` in the low level call?

It is also really weird to me to see `abi.encodePacked` with a dynamic sized parameter before a fixed sized one (instead of `abi.encode`) but I guess it should be fine because it is like that in the MinimalForwarder.sol code.

One last thing, I guess it will be fine to revert if there is an error in the low level call, right? There will be no need to batch this meta-calls, etc

Reply:
> So this is going to work just fine for whitelisting a requester but if this contract is going to be used for other purposes, shouldn't it also forward value in the low level call?

Imo it's fine for the same reason why value is excluded in our SelfMulticall/ExternalMulticall contracts

> It is also really weird to me to see abi.encodePacked with a dynamic sized parameter before a fixed sized one (instead of abi.encode) but I guess it should be fine because it is like that in the MinimalForwarder.sol code.
 
 That's an [EIP-2771](https://eips.ethereum.org/EIPS/eip-2771#extracting-the-transaction-signer-address) thing
 
 > One last thing, I guess it will be fine to revert if there is an error in the low level call, right? (This contract inherits SelfMulticall so batching call with the tryMulticall() function should be fine)
 
 Yes, that was my thinking. Maybe I should add a test for it.

Reply:
Added the test